streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+    stream_info = media_playlists[0].stream_info.copy()
+    
+    for key in list(stream_info):
+      for media_playlist in media_playlists:
+        # If a media playlist has a different value for this key, pop it.
+        if media_playlist.stream_info.get(key) != stream_info[key]:
+          stream_info.pop(key)
+          break
+    
+    return stream_info
+  
+  @staticmethod
+  def _get_bandwidth(var_playlists: List['MediaPlaylist'],
+                     durations: List[float]) -> Tuple[str, str]:
+    """A helper method to get the peak and average bandwidth
+    for stream variants.                     .---- dot product operator
+                                             |
+    BANDWIDTH = max((BANDWIDTH)s)            v
+    AVERAGE-BANDWIDTH = (AVERAGE-BANDWIDTH)s . (DURATION)s
+                        /sum((DURATION)s)
+    """
+    
+    band, avg_band = 0, 0.0
+    for var_playlist, duration in zip(var_playlists, durations):
+      band = max(int(var_playlist.stream_info['BANDWIDTH']), band)
+      avg_band += int(var_playlist.stream_info['AVERAGE-BANDWIDTH']) * duration
+    
+    return str(band), str(math.ceil(avg_band/sum(durations)))
+  
+  @staticmethod
+  def _get_codec(var_playlists: List['MediaPlaylist']) -> str:
+    """A helper method to get all the possible codecs for a concatenated stream
+    variant, which is all the codecs present in all the variants(in different
+    periods) that will be concatenated.
+    """
+    
+    codec_strings: Set[str] = set()
+    for var_playlist in var_playlists:
+      codec_strings.update(_unquote(
+          var_playlist.stream_info['CODECS']).split(','))
+    
+    return _quote(','.join(codec_string for
+                           codec_string in codec_strings))
+  
+  @staticmethod
+  def _next_unq_name() -> Tuple[str, str]:
+    """Returns a unique NAME and URI."""
+    
+    stream_name = 'stream_' + str(MediaPlaylist.current_stream_index)
+    MediaPlaylist.current_stream_index += 1
+    
+    return _quote(stream_name), _quote(stream_name + '.m3u8')
+  
+  @staticmethod
+  def _max_channels(aud_playlists: List['MediaPlaylist']
+                    ) -> str:
+    """Returns the maximum channel count in the given audio playlists."""
+    
+    return _quote(str(max(aud_playlist.channel_layout.max_channels for
+                          aud_playlist in aud_playlists)))
+  
+  @staticmethod
+  def _max_targer_dur(media_playlists: List['MediaPlaylist']) -> int:
+    """Returns the maximum target duration in the given media playlists."""
+    
+    return max(media_playlist.target_duration for
+               media_playlist in media_playlists)
+  
+  @staticmethod
+  def _fit_missing_lang(variant_options: List['MediaPlaylist'],
+                        language: str) -> str:
+    """Returns a substitution language for a missing language by considering
+    the languages of the given variants.
+    
+    Returns the argument `language` back only if no variants were given
```

Comment:
This is the argument passed to the method, it's like, if you don't provide me with variants i can choose form to replace the missing `language`, i will just return this missing language back to you.

This will only happen for when no text input in any language what so ever is present for some period, this means we will fill this period with the dummy subtitles for the period's duration to keep the subtitles of the next period(if there is) in sync.
This method will return the same missing language back and that will indicate that we failed to find a substitution for this language.

It's some fancy way to handle this though, i can just change it in the `concat_sub` method itself.